plugins/ipfs/local/localipfs.go

```diff
@@ -0,0 +1,504 @@
+		return nil, err
+	}
+
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+
+	err = cmd.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	pid := cmd.Process.Pid
+
+	err = ioutil.WriteFile(filepath.Join(dir, "daemon.pid"), []byte(fmt.Sprint(pid)), 0666)
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, ipfs.WaitOnAPI(l)
+}
+
+func (l *Localipfs) Stop(ctx context.Context, wait bool) error {
+	pid, err := l.getPID()
+	if err != nil {
+		return fmt.Errorf("error killing daemon %s: %s", l.dir, err)
+	}
+
+	p, err := os.FindProcess(pid)
+	if err != nil {
+		return fmt.Errorf("error killing daemon %s: %s", l.dir, err)
+	}
+
+	waitch := make(chan struct{}, 1)
+	go func() {
+		p.Wait() //TODO: pass return state
+		waitch <- struct{}{}
+	}()
+
+	defer func() {
+		err := os.Remove(filepath.Join(l.dir, "daemon.pid"))
+		if err != nil && !os.IsNotExist(err) {
+			panic(fmt.Errorf("error removing pid file for daemon at %s: %s", l.dir, err))
+		}
+	}()
+
+	if err := l.signalAndWait(p, waitch, syscall.SIGTERM, 1*time.Second); err != errTimeout {
+		return err
+	}
+
+	if err := l.signalAndWait(p, waitch, syscall.SIGTERM, 2*time.Second); err != errTimeout {
+		return err
+	}
+
+	if err := l.signalAndWait(p, waitch, syscall.SIGQUIT, 5*time.Second); err != errTimeout {
+		return err
+	}
+
+	if err := l.signalAndWait(p, waitch, syscall.SIGKILL, 5*time.Second); err != errTimeout {
+		return err
+	}
+
+	for {
+		err := p.Signal(syscall.Signal(0))
+		if err != nil {
+			break
+		}
+		time.Sleep(time.Millisecond * 10)
+	}
+
+	return nil
+}
+
+func (l *Localipfs) RunCmd(ctx context.Context, stdin io.Reader, args ...string) (testbedi.Output, error) {
+	env, err := l.env()
+
+	if err != nil {
+		return nil, fmt.Errorf("error getting env: %s", err)
+	}
+
+	cmd := exec.CommandContext(ctx, args[0], args[1:]...)
+	cmd.Env = env
+	cmd.Stdin = stdin
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	err = cmd.Start()
+
+	stderrbytes, err := ioutil.ReadAll(stderr)
+	if err != nil {
+		return nil, err
+	}
+
+	stdoutbytes, err := ioutil.ReadAll(stdout)
+	if err != nil {
+		return nil, err
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	exiterr := cmd.Wait()
+
+	var exitcode = 0
+	switch oerr := exiterr.(type) {
+	case *exec.ExitError:
+		if ctx.Err() == context.DeadlineExceeded {
+			err = errors.Wrapf(oerr, "context deadline exceeded for command: %q", strings.Join(cmd.Args, " "))
+		}
+
+		exitcode = 1
+	case nil:
+		err = oerr
+	}
+
+	return iptbutil.NewOutput(args, stdoutbytes, stderrbytes, exitcode, err), nil
+}
+
+func (l *Localipfs) Connect(ctx context.Context, tbn testbedi.Core) error {
+	swarmaddrs, err := tbn.SwarmAddrs()
+	if err != nil {
+		return err
+	}
+
+	_, err = l.RunCmd(ctx, nil, "ipfs", "swarm", "connect", swarmaddrs[0])
+
+	return err
+}
+
+func (l *Localipfs) Shell(ctx context.Context, nodes []testbedi.Core) error {
+	shell := os.Getenv("SHELL")
```

Comment:
would be nice if these env variables are global constants, such that other plugins could reuse them